### ISSUE TITLE
Custom bitmap targets and bitmap recycling

### ISSUE THREAD

**Issue by paveldudka** [bumptech/glide#1199]
I have a question regarding loading bitmaps into custom targets. I am wondering what is the right way of clearing resources.

I am loading bitmap this way:

``` java
Bitmap bmp = Glide.with(context)
    .load(resId)
    .asBitmap()
    .into(width, height)
    .get();
```

Once Im done using bitmap I received, is it a good idea to put it back into BitmapPool? Or it will be done automagically somewhere inside Glide? Logic tells me that's not gonna happen since Glide has no idea about lifecycle of that bitmap until something actually calls `Glide#clean(Target)` explicitly. So bitmap will eventually be garbage collected and will not find its way back into pool.

I believe ideally I need to hold onto target and clean it once Im done using bitmap, but in my case passing around target does not really look nice, so the only thing I have is a bitmap.

So the question is - can I just put used bitmap back into BitmapPool? Just want to make sure it will not break internal Glide assumptions about this bitmap and Glide will not try to put this bitmap back into pool too.

Also want to point out that `context` - is actually an Application context, so Glide will not get a hint regarding activity lifecycle

**Comment by TWiStErRob** [comment]
I was going to point you to http://stackoverflow.com/a/32887693/253468, but your last sentence cleared it up. Why do you need to pass the target around? Can't you do something like (as in the JavaDoc):
```java
FutureTarget<BItmap> target = Glide.with(context)
    .load(resId)
    .asBitmap()
    .into(width, height);
Bitmap bmp = target.get();
doStuff(bmp);
Glide.clear(target);
```
This must be called on a background thread anyway, because `get()` is blocking. If `doStuff` is async then you need to pass along the target, mind you that you could just pass around the `FutureTarget` instead of (not in addition to) the `Bitmap`, because once `get()` returns the second `get()` call is non-blocking:
```java
final FutureTarget<BItmap> target = Glide.with(context)
    .load(resId)
    .asBitmap()
    .into(width, height);
target.get(); // wait for it to load
runOnUiThread(() -> doStuff(target));

@UiThread
void doStuff(FutureTarget<Bitmap> target) {
    view.set(target.get());
    view.setTag(target);
}
```
and when you know you know you don't need it any more `Glide.clear((Target)view.getTag())`. The only problem is `RequestFutureTarget.assertBackgroundThread`, so I guess it requires to pass along both objects, or a simple two-field wrapper object.

If you describe exactly what you do with this Bitmap I can suggest the "Glide way" of doing it, if it's possible, using a real custom target.

@sjudd I think we could relax `doGet` to not assert background thread when `timeout` is null or 0 so this use case would work.

**Comment by sjudd** [comment]
@TWiStErRob Yes, I think this may have been done internally actually. 

The short answer to why you can't just return the Bitmap to the pool directly is that Glide uses reference counting, so others may have a reference to the same Bitmap.

**Comment by paveldudka** [comment]
I see. I think passing around FutureTarget instead of Bitmap is fine for me as well, so that's Im gonna go with I think.

Here is my use case:

I have custom model loader and to accommodate that I am using custom Decoder<MyModel, Bitmap>. The way it works - is my model describes 2 images and I need to compose these 2 images into bigger one:

```
class MyModel {
    int imgAResId;
    int imgBResId;
}
```
and my `Decoder#decode()` method roughly looks like this:
```
Resource<Bitmap> decode(MyModel model, int width, int height) {
    Bitmap a = Glide.with(context)
        .load(resId)
        .asBitmap()
        .into(width, height)
        .get();

    Bitmap b = Glide.with(context)
        .load(resId)
        .asBitmap()
        .into(width, height)
        .get();

    Bitmap result = compose2Bitmaps(a, b);
    return BitmapResource.obtain(bitmap, pool);
}
```

So as you can see, at the end I have 2 orphan bitmaps which I don't need anymore. I believe remembering FutureTarget and clearing it after result image is composed is a right way to go here.

Also I would appreciate comments regarding my approach in loading images inside Decoder.. Looks sketchy, but seemed like an easiest way to accomplish what Im trying to do.

**Comment by TWiStErRob** [comment]
Yep, that's the simplest case possible as displayed in the docs, you can easily transform that code into:
```java
FutureTarget<Bitmap> targetA = Glide...into;
FutureTarget<Bitmap> targetB = Glide...into;
try {
    Bitmap result = compose(targetA.get(), targetB.get(), width, height);
} finally {
    Glide.clear(targetA);
    Glide.clear(targetB);
}
return obtain(result);
```

If it works it's a good solution. The only drawback/edge case could be if the phone has only a few cores this may just deadlock on itself: this `decode` is trying to finish and is blocked on `.get()`, but in order to finish that `.get()` this `decode` has to finish to let the Glide thread pool pick up the task that trying to load `Bitmap a`; to fix this you just need to make sure to create the resize and disk cache services with at least 2 threads. Use `GlideBuilder` to set those: https://github.com/bumptech/glide/wiki/Configuration.

Actually thinking a little more it may be not enough to have just two threads, because even then it's still possible that two decodes start at the same time, but none of the `into(w,h).get()` will be able to complete, because all (2) pool threads are occupied. @sjudd what do you think, can this really happen? How can this be made safe?

Adding a timeout is an option, but then you wouldn't get a result or you'd have to retry.

Another for this would be to acquire both Bitmaps **before** starting the combination, something like below.
```java
class MyModel {
	private final int imgAResId;
	private final int imgBResId;
	private final int singleW;
	private final int singleH;
	MyModel(@AnyRes int imgAResId, @AnyRes int imgBResId, int singleW, int singleH) {
		this.imgAResId = imgAResId;
		this.imgBResId = imgBResId;
		this.singleW = singleW;
		this.singleH = singleH;
	}
	private AsyncTask<?, ?, ?> loadingTask;
	private FutureTarget<Bitmap> targetA;
	private FutureTarget<Bitmap> targetB;

	@WorkerThread
	Bitmap getBitmapA(RequestManager glide) throws ExecutionException, InterruptedException {
		if (targetA == null) {
			targetA = glide.load(imgAResId).asBitmap().into(singleW, singleH);
		}
		return targetA.get();
	}
	@UiThread
	void cancel() {
		if (loadingTask != null) {
			loadingTask.cancel(true);
		}
	}
	void clear() {
		if (targetA != null) {
			Glide.clear(targetA);
		}
	}
	@UiThread
	void loadInto(final RequestManager glide, final ImageView imageView) {
		loadingTask = new AsyncTask<Object, Object, Object>() {
			@Override protected Object doInBackground(Object[] params) {
				try {
					// make sure both Bitmaps are readily available ...
					getBitmapA(glide);
					getBitmapB(glide);
				} catch (ExecutionException | InterruptedException ex) {
					// TODO
				}
				return null;
			}
			@Override protected void onPostExecute(Object o) {
				if (!isCancelled()) {
					// ... before actually starting the decoder
					glide.load(this).into(imageView);
				}
			}
		};
	}
}
class ViewHolder {
	MyModel current;
	ImageView image;
	void bind(RequestManager glide, MyModel model) {
		if (current != null) {
			current.cancel();
		}
		current = model;
		current.loadInto(glide, image);
	}
}
Resource<Bitmap> decode(MyModel model, int width, int height) {
	Bitmap result = compose2Bitmaps(model.getBitmapA(), model.getBitmapB());
	model.clear();
	return obtain(result);
}
```
Sorry for the mess, it's a first draft to get the idea through. This may be an overkill depending on your situation. The ViewHolder and cancellation is needed only if you use these models in a list, because if an item is recycled the preparation has to be stopped.

**Comment by paveldudka** [comment]
Yeah.. Just tried to run 8 simultaneous loads and it does dead lock..
Also having a pre-load method which accepts ImageView does not really work for me because this logic is shared between different targets. In some cases it is an ImageView, in some cases it is just a synchronous call with FutureTarget (when I need a final bitmap to display in Notification f.i.) Obviously I can work around that by passing some sort of callback instead of ImageView and let caller decide what to do with the result or even pass a Target, but this approach indeed looks messy :( Especially when it comes to animating the result

If I wont find a better solution, will go back to this solution though

**Comment by paveldudka** [comment]
So as far as I get the overall architecture, is it correct to assume that I need `num_of_parallel_decode_calls + 1` threads in order to make it work?

**Comment by paveldudka** [comment]
I wish I could specify ExecutorService exclusively for those inner image requests..

**Comment by TWiStErRob** [comment]
Hmm, how about taking a step back: since your input is just two local resources anyway, you could just write a decoder which simply uses `res.getDrawable()` and draw those two onto a Canvas backed by a Bitmap. This would greatly simplify things on the account that you'd need a little bit more complicated combine function which does some resizing while drawing if needed.

**Comment by TWiStErRob** [comment]
Yes, that thread count seems correct, I had a similar feeling.

As for the notification, you can modify my code to accept a `Target<Bitmap>` and call it with `new BitmapImageViewTarget(holder.imageView)` and `new SimpleTarget () {...}` for the notification, or even use the builtin NotificationTarget if that works for you.

**Comment by TWiStErRob** [comment]
For getting singlew/h you can ask the target passed in via `getSize(new ...)` and start the async task in there.

**Comment by paveldudka** [comment]
My input is not just 2 local resources. My example is just a simplified version of more complex logic of "preparing" single bitmap. Essentially here is what it looks like:

```
class MyModel {
    SingleModel a;
    SingleModel b;
}

class SingleModel {
    URI uri;
    String text;
}

Bitmap decodeSingleModel(SingleModel model, int w, int h) {
    Bitmap result = .....
    if(model.uri != null) {
        result = ... //load from URI using Glide
    } else if(text != null) {
       result = ... //manually render text into bitmap
    } else {
       result = ...//fall back to local resourceId using Glide
    }
    return result;
}


Bitmap renderFinalBitmap(Bitmap a, Bitmap b, int w, int h) {
    // combine 2 smaller bitmaps together
}
```

**Comment by TWiStErRob** [comment]
Ah, I see ... it all makes sense now :)

Forgot to say: check out the related #1037 and the linked SO answer.

Also try to play with `.priority()` to prevent deadlock, I think the single model loads should have higher priority.

**Comment by paveldudka** [comment]
Thanks a lot for your prompt responses! Have plenty of options to play with now :)

**Comment by paveldudka** [comment]
One thing I don't really understand though is even after I changed ExecutorService to the "growing" one (i.e. bumped up `maxPoolSize` value to a higher number), Glide is still creating `corePoolSize` number of threads and then blocks.

I suspect Im missing some fundamental Java concept here though

**Comment by TWiStErRob** [comment]
There are two services, are you sure you replaced the right one?

**Comment by paveldudka** [comment]
Yes, there are disk cache service and `resize` service which is used not only by a resize :). So I specified custom `resize` service. Also specified custom thread factory and added log whenever new thread is created, so I can see it. Core pool size - that's when new threads stop coming out

**Comment by TWiStErRob** [comment]
Can you share the relevant GlideModule lines?

**Comment by paveldudka** [comment]
```
@Override public void applyOptions(Context context, GlideBuilder builder) {
        builder.setResizeService(
                new FifoPriorityThreadPoolExecutor(2,
                        20,
                        0,
                        TimeUnit.MILLISECONDS,
                        new ThreadFactory() {
                            int threadNum = 0;

                            @Override
                            public Thread newThread(@NonNull Runnable runnable) {
                                final Thread result = new Thread(runnable, "my-service-thread-" + threadNum) {
                                    @Override
                                    public void run() {
                                        android.os.Process.setThreadPriority(android.os.Process.THREAD_PRIORITY_BACKGROUND);
                                        super.run();
                                    }
                                };
                                threadNum++;
                                Log.e("test", "created thread: " + result.getName());
                                return result;
                            }
                        },
                        FifoPriorityThreadPoolExecutor.UncaughtThrowableStrategy.LOG));

    }
```

**Comment by TWiStErRob** [comment]
Hmm, I couldn't find anything weird just be reading code.
Is it possible that your load timings work out so that you actually don't need more than `core` threads?
(I'm out for now, it's 1am)

**Comment by TWiStErRob** [comment]
Also the default diskCacheStrategy is RESULT, so some of the image(s) are possible loaded in the other pool. Make sure you do `.diskCacheStrategy(NONE).skipMemoryCache(true)` on all related loads while testing.

**Comment by paveldudka** [comment]
Not really - my load deadlocked so it definitely wants more threads.

**Comment by paveldudka** [comment]
Ill give it a try. Go to sleep!

**Comment by paveldudka** [comment]
It turns out it has something to do with `FifoPriorityThreadPoolExecutor`. Changing it to `Executors.newCachedThreadPool` solves the issue - pool now grows as new simultaneous tasks come in.

Still not quite sure it is a good idea to let it grow with that pace... Need to run profiling to see how does really affect app performance

**Comment by TWiStErRob** [comment]
If you mean instrumented profiling, be careful with Glide.load lines, they show up really skewed in the timeline because there're a lot of small methods, it's much faster than advertised there. (This was my experience on Dalvik at least, not sure how Art result look like yet.)

**Comment by paveldudka** [comment]
No. The log line from my thread factory. Now I can see new threads are being created beyond corePoolSize point.

I actually even created a small app with FifoPriorityThreadPoolExecutor - submit new runnable to it once button is clicked. And I got the same problem - executer stops creating new threads once corePoolSize number of threads is reached. mapPoolSize is ignored

**Comment by paveldudka** [comment]
I think this issue can be closed as it drifted already from the original problem.

If you are interested, I did some research why specifying maxPoolSize doesnt make thread pool grow when number of active threads reaches corePoolSize value. And the reason is behind using BlockingPriorityQueue used inside FifoPriorityThreadPoolExecutor. 

The way ThreadPoolExecutor decides if we need new thread or not is the following:

1. If number of threads < corePoolSize - create new thread
2. Else try to put task in the queue - if succeeded - fine. Otherwise create new thread

The problem is that BlockingPriorityQueue always allows you to add task to the queue (as per documentation), so ThreadPoolExecutor never creates new threads once number of threads reaches corePoolSize. All subsequent tasks are just placed into queue.

However, I didn't like a simple growing thread pool as I'm my case I might have 10-15 simultaneous glide requests at the same time (which in turn will produce 10-15 more threads for inner requests), so I created a simple wrapper for ExetutorService which redirects tasks with Priority.IMMEDIATE to a dedicated thread pool, so I don't have a deadlock
